File: Kazak-Android/app/src/main/java/io/kazak/notifications/EventAlarmService.java
package io.kazak.notifications;

import android.app.IntentService;
import android.app.Notification;
import android.content.Intent;
import android.util.Log;

import javax.inject.Inject;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

import io.kazak.KazakApplication;
import io.kazak.model.Day;
import io.kazak.model.Schedule;
import io.kazak.model.Talk;
import io.kazak.repository.DataRepository;
import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;

public class EventAlarmService extends IntentService {

    private static final int NOTIFICATION_INTERVAL_MINUTES = 10;

    @Inject
    @SuppressWarnings("checkstyle:visibilitymodifier")
    DataRepository dataRepository;

    public EventAlarmService() {
        super(EventAlarmService.class.getName());
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        KazakApplication.injector().inject(this);

        final NotificationCreator notificationCreator = new NotificationCreator(this);
        final Notifier notifier = Notifier.from(this);

        Calendar calendar = Calendar.getInstance();
        final Date now = calendar.getTime();
        calendar.add(Calendar.MINUTE, NOTIFICATION_INTERVAL_MINUTES);
        final Date notificationInterval = calendar.getTime();

        Observable<Talk> talks = loadTalks();

        talks.filter(startingIn(now, notificationInterval))
                .toList()
                .map(createNotifications(notificationCreator))
                .subscribe(displayNotifications(notifier));

    }

    private Observable<Talk> loadTalks() {
        return dataRepository
                .getSchedule()
                .take(1)
                .flatMap(getSortedDays())
                .flatMap(getSortedTalks());
    }

    private Func1<Schedule, Observable<Day>> getSortedDays() {
        return new Func1<Schedule, Observable<Day>>() {
            @Override
            public Observable<Day> call(Schedule schedule) {
                List<Day> days = schedule.getDays();
                Collections.sort(
                        days, new Comparator<Day>() {
                            @Override
                            public int compare(Day firstDay, Day secondDay) {
                                return firstDay.getDay().compareTo(secondDay.getDay());
                            }
                        }
                );
                return Observable.from(days);
            }
        };
    }

    private Func1<Day, Observable<Talk>> getSortedTalks() {
        return new Func1<Day, Observable<Talk>>() {
            @Override
            public Observable<Talk> call(Day day) {
                List<Talk> talks = day.getTalks();
                Collections.sort(
                        talks, new Comparator<Talk>() {
                            @Override
                            public int compare(Talk firstTalk, Talk secondTalk) {
                                Date firstTalkStart = firstTalk.getTimeSlot().getStart();
                                Date secondTalkStart = secondTalk.getTimeSlot().getStart();
                                return firstTalkStart.compareTo(secondTalkStart);
                            }
                        }
                );
                return Observable.from(talks);
            }
        };
    }

    private Func1<Talk, Boolean> startingIn(final Date now, final Date notificationInterval) {
        return new Func1<Talk, Boolean>() {
            @Override
            public Boolean call(Talk talk) {
                Date talkStart = talk.getTimeSlot().getStart();
                return now.before(talkStart) && (talkStart.before(notificationInterval) || talkStart.equals(notificationInterval));
            }
        };
    }

    private Func1<List<Talk>, List<Notification>> createNotifications(final NotificationCreator notificationCreator) {
        return new Func1<List<Talk>, List<Notification>>() {
            @Override
            public List<Notification> call(List<Talk> talks) {
                return notificationCreator.createFrom(talks);
            }
        };
    }

    private Action1<List<Notification>> displayNotifications(final Notifier notifier) {
        return new Action1<List<Notification>>() {
            @Override
            public void call(List<Notification> notifications) {
                notifier.showNotifications(notifications);
            }
        };
    }

}

File: Kazak-Android/app/src/main/kotlin/io/kazak/notifications/EventAlarmService.kt
package io.kazak.notifications

import android.app.AlarmManager
import android.app.IntentService
import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import io.kazak.KazakApplication
import io.kazak.model.Session
import io.kazak.repository.DataRepository
import rx.Observable
import java.util.Calendar
import java.util.Date

class EventAlarmService : IntentService(EventAlarmService::class.java.name) {

    companion object {
        private val NOTIFICATION_INTERVAL_MINUTES = 10
    }

    val dataRepository: DataRepository

    init {
        dataRepository = KazakApplication.injector().dataRepository
    }

    override fun onHandleIntent(intent: Intent) {
        val notificationCreator = NotificationCreator(this)
        val notifier = Notifier.from(this)

        val calendar = Calendar.getInstance()
        val now = calendar.time
        calendar.add(Calendar.MINUTE, NOTIFICATION_INTERVAL_MINUTES)
        val notificationInterval = calendar.time

        val sessions = dataRepository.getFavorites()
                .take(1)
                .map { sortByStartingTime(it) }
                .flatMap { Observable.from(it) }

        sessions.filter { startingIn(now, notificationInterval, it) }
                .toList()
                .map { createNotifications(notificationCreator, it) }
                .subscribe { displayNotifications(it, notifier) }

        sessions.filter { startingAfter(notificationInterval, it) }
                .take(1)
                .subscribe { scheduleNextAlarm(it) }
    }

    private fun sortByStartingTime(sessions: List<Session>): List<Session> {
        return sessions.sortedBy { it.timeSlot().start }
    }

    private fun startingIn(now: Date, notificationInterval: Date, session: Session): Boolean {
        val sessionStart = session.timeSlot().start
        return now.before(sessionStart) && (sessionStart.before(notificationInterval) || sessionStart == notificationInterval)
    }

    private fun createNotifications(notificationCreator: NotificationCreator, sessions: List<Session>): List<Notification> {
        return notificationCreator.createFrom(sessions)
    }

    private fun displayNotifications(notifications: List<Notification>, notifier: Notifier) {
        notifier.showNotifications(notifications)
    }

    private fun startingAfter(notificationInterval: Date, session: Session): Boolean {
        val sessionStart = session.timeSlot().start
        return sessionStart.after(notificationInterval)
    }

    private fun scheduleNextAlarm(session: Session) {
        val talkStart = session.timeSlot().start
        val calendar = Calendar.getInstance()
        calendar.time = talkStart
        calendar.add(Calendar.MINUTE, -NOTIFICATION_INTERVAL_MINUTES)

        val serviceIntent = Intent(this@EventAlarmService, EventAlarmService::class.java)
        val pendingIntent = PendingIntent.getService(this@EventAlarmService, 0, serviceIntent, PendingIntent.FLAG_CANCEL_CURRENT)
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, pendingIntent)
    }

}
